File: python/thomson_multitaper/multitaper_adaptive.py
"""Translation of MultitaperAdaptive.m: exact multitaper spectral estimator
with Thomson's adaptive (data-driven) taper weighting. Original Copyright
2019, Santhosh Karnik.

Deviations from the .m source (see NOTES.md):
  - The original's `cutoff > eps` branch calls `median(1, ceil(...)+10, N)`
    (missing the `[...]` that Multitaper.m uses around the same expression),
    which in MATLAB would error or misbehave rather than compute a median of
    three values. Translated here as the evident intent: `median([1, ..., N])`.
  - The original does not clamp K to be at least 1 (Multitaper.m does, via
    `K = max(K,1)`); clamped here too for robustness, since an all-zero
    taper set would otherwise crash spectral_estimate.
"""
from __future__ import annotations

import numpy as np

from ._utils import datawrap, dpss


class MultitaperAdaptive:
    def __init__(self, n: int, w: float, cutoff: float):
        self.n = n
        self.w = w
        self.cutoff = cutoff
        eps = np.finfo(float).eps

        if cutoff >= 1:
            k = int(np.floor(cutoff + eps))
            s, lam = dpss(n, n * w, k)
        elif cutoff > eps:
            if cutoff > 0.5:
                k_est = int(np.ceil(2 * n * w))
            else:
                k_est = int(np.median([1, np.ceil(2 * n * w + (2 / np.pi**2) * np.log(n) * np.log(1.0 / cutoff)) + 10, n]))
            s_full, lam_full = dpss(n, n * w, k_est)
            k = max(int(np.count_nonzero(lam_full > cutoff)), 1)
            s, lam = s_full[:, :k], lam_full[:k]
        else:
            raise ValueError("cutoff must be >= 1 or > eps")

        self.S = s
        self.lambda_ = lam
        self.K = k

    def spectral_estimate(self, x: np.ndarray, m: int | None = None):
        """Adaptively-weighted multitaper PSD estimate.

        Returns (Sf, Skf, wkf, num_iter): the combined estimate, the
        per-taper single-taper estimates, the final weights, and the
        iteration count -- matching the .m source's four outputs.
        """
        if m is None:
            m = self.n
        x = np.asarray(x)

        if self.n > m:
            sx = self.S * x[:, None]
            skf = np.stack([datawrap(sx[:, k], m) for k in range(self.K)], axis=1)
        else:
            skf = self.S * x[:, None]

        skf = np.fft.fft(skf, n=m, axis=0)
        skf = skf.real**2 + skf.imag**2

        sigma2 = float(np.vdot(x, x).real) / self.n
        oneminuslambdasigma2 = np.tile((1 - self.lambda_) * sigma2, (m, 1))

        wkf = np.tile(self.lambda_, (m, 1))
        sf = (skf[:, 0] + skf[:, 1]) / 2
        sf_old = np.zeros(m)
        num_iter = 0
        while np.sum(np.abs(sf - sf_old)) > 5e-4 * sigma2 / m:
            num_iter += 1
            sf_old = sf
            wkf = (sf[:, None] ** 2 * self.lambda_[None, :]) / (
                sf[:, None] * self.lambda_[None, :] + oneminuslambdasigma2
            ) ** 2
            sf = np.sum(wkf * skf, axis=1) / np.sum(wkf, axis=1)

        return sf, skf, wkf, num_iter

File: legacy/matlab_source/ThomsonsMethodRevisitedExperiments/MultitaperAdaptive.m
function MTA = MultitaperAdaptive(N,W,cutoff)
%% MultitaperAdaptive.m 
% Function implementing the exact multitaper spectral estimator with an adaptive weighting scheme
% 
% Inputs:
% -N: Length of DPSSs
% -W: Half Bandwidth for DPSSs
% -cutoff: Projection onto the first K = cutoff DPSSs 
%          OR Projection onto the first K = #{ell : lambda_{N,W}^{(ell)} >= cutoff} DPSSs.
%
% Outputs:
% -MTA an object containing an exact Multitaper Spectral Estimate function with adaptive weights
%
% Given an Nx1 vector x, performing
% y = MTA.SpectralEstimate(x,M);
% results in y[m] as the adaptively weighted multitaper spectral estimate at f = m/M
%
% Notes: 
%
% Most recent change - 01/30/2019
%
% Copyright 2019, Santhosh Karnik

%% Multitaper Adaptive Function
MTA.SpectralEstimate = @SpectralEstimate;

%% Store signal length, bandwidth, and number of DPSSs
MTA.N = N;
MTA.W = W;
MTA.cutoff = cutoff;

%% Compute and store DPSSs
if(cutoff >= 1)
    K = floor(cutoff+eps);
    [MTA.S,MTA.lambda] = dpss(N,N*W,K);
elseif(cutoff > eps)
    if(cutoff > 0.5)
        Kest = ceil(2*N*W);
    else
        Kest = median(1,ceil(2*N*W+(2/pi^2)*log(N)*log(1./cutoff))+10,N);
    end
    [S,lambda] = dpss(N,N*W,Kest);
    K = nnz(lambda > cutoff);
    MTA.S = S(:,1:K);
    MTA.lambda = lambda(1:K);
end
MTA.K = K;

%% Multitaper Spectral Estimation with Adaptive Weights
function [Sf,Skf,wkf,num_iter] = SpectralEstimate(x,M)
    if(nargin < 2)
        M = MTA.N;
    end
    
    % Compute single taper estimates for each DPSS taper
    if(MTA.N > M)
       Sx = bsxfun(@times,MTA.S,x);
       % Datawrap Sx
       Skf = zeros(M,MTA.K);
       for k = 1:MTA.K,
           Skf(:,k) = datawrap(Sx(:,k),M);
       end
    else
        Skf = bsxfun(@times,MTA.S,x);
    end    
    Skf = fft(Skf,M,1);
    Skf = real(Skf).^2+imag(Skf).^2;
    
    % Iterative scheme to compute frequency dependent weights and weighted spectral estimate
    sigma2 = x'*x/MTA.N;
    oneminuslambdasigma2 = repmat((1-MTA.lambda')*sigma2,M,1);
    
    wkf = repmat(MTA.lambda',M,1);
    Sf = (Skf(:,1)+Skf(:,2))/2;
    Sf_old = zeros(M,1);
    num_iter = 0;
    while(sum(abs(Sf-Sf_old)) > 5e-4*sigma2/M)
        num_iter = num_iter + 1;
        Sf_old = Sf;
        wkf = ((Sf.^2)*MTA.lambda')./(Sf*MTA.lambda'+oneminuslambdasigma2).^2;
        Sf = sum(wkf.*Skf,2)./sum(wkf,2);
    end
    
end

end